File: day2.py
with open("inputs/day2.txt", "r") as f:
    inputLines = f.readlines()

def everyLine(line):
    reports = [int(x) for x in line.split()]
    previousReport = None
    increase = None
    safe = True
    safeDampened = 0
    for report in reports:
        if increase == None:
            if previousReport:
                if previousReport > report: 
                    increase = False
                elif previousReport == report:
                    safe = False
                    safeDampened += 1
                else: 
                    increase = True
        if previousReport and (increase != None):
            if (previousReport > report and increase == True) or (previousReport < report and increase == False) or (previousReport == report) or (increase and (previousReport <= report - 4)) or (not increase) and (previousReport >= report + 4):
                safe = False
        previousReport = report
    return safe

def partOne():
    safeTotal = 0
    for line in inputLines:
        safeTotal += everyLine(line)
    print(safeTotal)

def partTwo():
    safeTotal = 0
    for line in inputLines:
        if everyLine(line):
            safeTotal += 1
            continue
        for i in range(len(line)):
            if everyLine(line[:i-1] + line[i+1:]):
                safeTotal += 1
                break
    print(safeTotal)



partOne()
partTwo()

File: day2.mjs
import { readFile } from "node:fs/promises";

const input = await readFile("inputs/day2.txt", "utf8");
const lines = input.split("\r\n");

function everyLine(line) {
  const reports = line
    .split(" ")
    .filter((x) => x.trim() !== "")
    .map(Number);
  let previousReport = null;
  let increase = null;
  let safe = true;
  let safeDampened = 0;

  for (const report of reports) {
    if (increase === null) {
      if (previousReport !== null) {
        if (previousReport > report) {
          increase = false;
        } else if (previousReport === report) {
          safe = false;
          safeDampened++;
        } else {
          increase = true;
        }
      }
    }

    if (previousReport !== null && increase !== null) {
      if (
        (previousReport > report && increase === true) ||
        (previousReport < report && increase === false) ||
        previousReport === report ||
        (increase && previousReport <= report - 4) ||
        (!increase && previousReport >= report + 4)
      ) {
        safe = false;
      }
    }

    previousReport = report;
  }

  return safe;
}

function partOne() {
  let safeTotal = 0;
  for (const line of lines) {
    if (everyLine(line)) {
      safeTotal++;
    }
  }
  console.log(safeTotal);
}

function partTwo() {
  let safeTotal = 0;
  for (const line of lines) {
    if (everyLine(line)) {
      safeTotal++;
      continue;
    }
    for (let i = 0; i < line.split(" ").length; i++) {
      const modifiedLine = line
        .split(" ")
        .filter((_, index) => index !== i)
        .join(" ");
      if (everyLine(modifiedLine)) {
        safeTotal++;
        break;
      }
    }
  }
  console.log(safeTotal);
}

partOne();
partTwo();
